main.tf

```diff
@@ -109,6 +109,12 @@ resource "aws_dms_replication_subnet_group" "this" {
   subnet_ids                           = var.repl_subnet_group_subnet_ids
 
   tags = merge(var.tags, var.repl_subnet_group_tags)
+
+  depends_on = [
+    aws_iam_role.dms_access_for_endpoint,
+    aws_iam_role.dms_cloudwatch_logs_role,
+    aws_iam_role.dms_vpc_role,
```

Comment:
looks like we only need to depend on the `aws_iam_role.dms_vpc_role`. I tested with this small repro:

```hcl
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.0"

  name = "test"
  cidr = "10.99.0.0/18"

  azs              = ["us-east-1a", "us-east-1b", "us-east-1d"]
  public_subnets   = ["10.99.0.0/24", "10.99.1.0/24", "10.99.2.0/24"]
  private_subnets  = ["10.99.3.0/24", "10.99.4.0/24", "10.99.5.0/24"]
  database_subnets = ["10.99.7.0/24", "10.99.8.0/24", "10.99.9.0/24"]

  create_database_subnet_group = true
  enable_nat_gateway           = false
  single_nat_gateway           = true
}

resource "aws_dms_replication_subnet_group" "this" {
  replication_subnet_group_id          = "test"
  replication_subnet_group_description = "test"
  subnet_ids                           = module.vpc.database_subnets

  depends_on = [
    aws_iam_role.dms_vpc_role
  ]
}

data "aws_iam_policy_document" "dms_assume_role" {
  statement {
    actions = ["sts:AssumeRole"]

    principals {
      identifiers = ["dms.amazonaws.com"]
      type        = "Service"
    }
  }
}

resource "aws_iam_role" "dms_vpc_role" {
  name                  = "dms-vpc-role"
  description           = "DMS IAM role for VPC permissions"
  assume_role_policy    = data.aws_iam_policy_document.dms_assume_role.json
  managed_policy_arns   = ["arn:aws:iam::aws:policy/service-role/AmazonDMSVPCManagementRole"]
  force_detach_policies = true

  # https://github.com/hashicorp/terraform-provider-aws/issues/11025#issuecomment-660059684
  provisioner "local-exec" {
    command = "sleep 10"
  }
}
```

I tested without the role and can confirm the subnet fails due to insufficient role - very odd but 🤷🏽